subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/AvailableJavaHomes.java

```diff
@@ -190,113 +199,102 @@ public static File getBestJre() {
     }
 
     private static List<JvmInstallation> getJvms() {
-        if (jvms == null) {
-            NativeServices nativeServices = NativeServicesTestFixture.getInstance();
-            FileCanonicalizer fileCanonicalizer = nativeServices.get(FileCanonicalizer.class);
-            jvms = new ArrayList<>();
-            jvms.addAll(new DevInfrastructureJvmLocator(fileCanonicalizer).findJvms());
-            InstalledJvmLocator installedJvmLocator = new InstalledJvmLocator(OperatingSystem.current(), Jvm.current(), nativeServices.get(WindowsRegistry.class), nativeServices.get(SystemInfo.class), fileCanonicalizer);
-            jvms.addAll(new DevInfrastructureJvmLocator(fileCanonicalizer).findJvms());
-            jvms.addAll(installedJvmLocator.findJvms());
-            if (OperatingSystem.current().isLinux()) {
-                jvms.addAll(new BaseDirJvmLocator(fileCanonicalizer, new File("/opt")).findJvms());
-            }
-            if (!OperatingSystem.current().isWindows()) {
-                jvms.addAll(new SdkManJvmLocator(fileCanonicalizer).findJvms());
-            }
-            jvms.addAll(new HomeDirJvmLocator(fileCanonicalizer).findJvms());
-            jvms.addAll(new EnvVariableJvmLocator().findJvms());
-            // Order from most recent to least recent
-            Collections.sort(jvms, (o1, o2) -> o2.getVersion().compareTo(o1.getVersion()));
-        }
+        return jvms.get();
+    }
+
+    private static List<JvmInstallation> discoverLocalInstallations() {
+        ExecHandleFactory execHandleFactory = TestFiles.execHandleFactory();
+        JvmVersionDetector versionDetector = new CachingJvmVersionDetector(new DefaultJvmVersionDetector(execHandleFactory));
+        final List<JvmInstallation> jvms = new SharedJavaInstallationRegistry(defaultInstallationSuppliers())
+            .listInstallations().stream()
+            .map(i -> asJvmInstallation(i, versionDetector))
+            .sorted(Comparator.comparing(JvmInstallation::getJavaVersion))
+            .collect(Collectors.toList());
+
         System.out.println("Found the following JVMs:");
         for (JvmInstallation jvm : jvms) {
             System.out.println("    " + jvm);
         }
         return jvms;
     }
 
+    private static List<InstallationSupplier> defaultInstallationSuppliers() {
+        WindowsRegistry windowsRegistry = NativeServicesTestFixture.getInstance().get(WindowsRegistry.class);
+        return Lists.newArrayList(
+            new AsdfInstallationSupplier(providerFactory()),
+            new BaseDirJvmLocator("/opt"),
+            new BaseDirJvmLocator(SystemProperties.getInstance().getUserHome()),
+            new CurrentInstallationSupplier(providerFactory()),
+            new DevInfrastructureJvmLocator(),
+            new EnvVariableJvmLocator(),
+            new JabbaInstallationSupplier(providerFactory()),
+            new LinuxInstallationSupplier(providerFactory()),
```

Comment:
Should we filter some of these since we know the OS we are on?

Reply:
No need, the suppliers will take care of that. I decided to make it a responsibility of the suppliers so consumers (and our injection framework) don't need to conditionally provide certain suppliers